core/src/main/scala/org/apache/spark/NewAccumulator.scala

```diff
@@ -0,0 +1,356 @@
+   */
+  def newId(): Long = nextId.getAndIncrement
+
+  /**
+   * Register an [[Accumulator]] created on the driver such that it can be used on the executors.
+   *
+   * All accumulators registered here can later be used as a container for accumulating partial
+   * values across multiple tasks. This is what [[org.apache.spark.scheduler.DAGScheduler]] does.
+   * Note: if an accumulator is registered here, it should also be registered with the active
+   * context cleaner for cleanup so as to avoid memory leaks.
+   *
+   * If an [[Accumulator]] with the same ID was already registered, this does nothing instead
+   * of overwriting it. We will never register same accumulator twice, this is just a sanity check.
+   */
+  def register(a: NewAccumulator[_, _]): Unit = synchronized {
+    if (!originals.containsKey(a.id)) {
+      originals.put(a.id, new jl.ref.WeakReference[NewAccumulator[_, _]](a))
+    }
+  }
+
+  /**
+   * Unregister the [[Accumulator]] with the given ID, if any.
+   */
+  def remove(id: Long): Unit = synchronized {
+    originals.remove(id)
+  }
+
+  /**
+   * Return the [[Accumulator]] registered with the given ID, if any.
+   */
+  def get(id: Long): Option[NewAccumulator[_, _]] = synchronized {
+    Option(originals.get(id)).map { ref =>
+      // Since we are storing weak references, we must check whether the underlying data is valid.
+      val acc = ref.get
+      if (acc eq null) {
+        throw new IllegalAccessError(s"Attempted to access garbage collected accumulator $id")
+      }
+      acc
+    }
+  }
+
+  /**
+   * Clear all registered [[Accumulator]]s. For testing only.
+   */
+  def clear(): Unit = synchronized {
+    originals.clear()
+  }
+}
+
+
+case class UpdatedLongValue(l: Long) extends UpdatedValue {
+  override def toString: String = l.toString
+}
+
+class LongAccumulator extends NewAccumulator[jl.Long, jl.Long] {
+  @transient private[this] var _sum = 0L
+
+  override def updatedValue: UpdatedValue = new UpdatedLongValue(_sum)
+
+  override def add(v: jl.Long): Unit = _sum += v
+
+  private[spark] def unboxAdd(v: Long): Unit = _sum += v
+
+  private[spark] def unboxValue: Long = _sum
+
+  private[spark] def setValue(newValue: Long): Unit = _sum = newValue
+
+  override def isNoOp(updates: UpdatedValue): Boolean = {
+    val v = updates.asInstanceOf[UpdatedLongValue].l
+    v == 0
+  }
+
+  override def applyUpdates(updates: UpdatedValue): Unit = {
+    val v = updates.asInstanceOf[UpdatedLongValue].l
+    _sum += v
+  }
+
+  override def localValue: jl.Long = _sum
+}
+
+
+case class UpdatedDoubleValue(d: Double) extends UpdatedValue {
+  override def toString: String = d.toString
+}
+
+class DoubleAccumulator extends NewAccumulator[jl.Double, jl.Double] {
+  @transient private[this] var _sum = 0.0
+
+  override def updatedValue: UpdatedValue = new UpdatedDoubleValue(_sum)
+
+  override def add(v: jl.Double): Unit = _sum += v
+
+  private[spark] def unboxAdd(v: Double): Unit = _sum += v
```

Comment:
i'd expose the specialized methods as public APIs, and just have them name the same thing as the generic methods so all users will automatically benefit from them.